### PR TITLE
Added tests for a Book functionality, fixed previous bugs

### PR DESCRIPTION
**Pull request by citruscaviar**
Book_Category_Test_Coverage

### CODE REVIEW

File: src/main/java/com/onlinebookstore/store/controller/CategoryController.java [hunk 1]
```diff
@@ -50,7 +49,7 @@ public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(@PathVariable Long c
     @GetMapping
     @Operation(description = "Finds all categories")
     @PreAuthorize("hasRole('USER') or hasRole('ADMIN')")
-    public Page<CategoryResponseDto> getAll(Pageable pageable) {
+    public List<CategoryResponseDto> getAll(Pageable pageable) {
```
Comment (JJJazl) [inline src/main/java/com/onlinebookstore/store/controller/CategoryController.java:52]: Why did you change it?

File: src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@Sql(scripts = {
+        "classpath:database/test/books/add-books-to-table.sql",
+        "classpath:database/test/categories/add-categories-to-table.sql",
+        "classpath:database/test/books/add-books-and-categories-into-table.sql"
+}, executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+@Sql(scripts = {
+        "classpath:database/test/books/delete-books-categories.sql",
+        "classpath:database/test/books/remove-books-from-table-books.sql",
+        "classpath:database/test/categories/delete-categories.sql"
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java:41]:
```suggestion
        "classpath:database/test/books/remove-books-categories.sql",
        "classpath:database/test/books/remove-books-from-table-books.sql",
        "classpath:database/test/categories/remove-categories.sql"
```

File: src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@Sql(scripts = {
+        "classpath:database/test/books/add-books-to-table.sql",
+        "classpath:database/test/categories/add-categories-to-table.sql",
+        "classpath:database/test/books/add-books-and-categories-into-table.sql"
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java:36]:
```suggestion
        "classpath:database/test/books/add-books.sql",
        "classpath:database/test/categories/add-categories.sql",
        "classpath:database/test/books/add-books-and-categories-relations.sql"
```

File: src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                                .content(jsonRequest)
+                                .contentType(MediaType.APPLICATION_JSON)
+                )
+                .andExpect(status().isOk())
+                .andReturn();
+
+        BookResponseDto actualDto = objectMapper.readValue(result.getResponse()
+                        .getContentAsString(),
+                BookResponseDto.class);
+        Assertions.assertEquals(expectedDto, actualDto);
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java:81]: Use static import for this method

File: src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    @DisplayName("Given valid book ID and update request, should return updated book")
+    void updateByValidId_ShouldReturnUpdatedBook_Ok() throws Exception {
+        long bookId = 12L;
+        BookRequestDto updateRequestDto = createUpdateRequestDto();
+        BookResponseDto expectedDto = createExpectedBookDto(bookId, updateRequestDto);
+
+        String jsonRequest = objectMapper.writeValueAsString(updateRequestDto);
+
+        MvcResult result = mockMvc.perform(
+                        put("/books/" + bookId)
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java:71]:
```suggestion
                        put("/books/{id}", bookId)
```

File: src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+                .webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+    }
+
+    @Test
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
+    @DisplayName("Given valid book ID and update request, should return updated book")
+    void updateByValidId_ShouldReturnUpdatedBook_Ok() throws Exception {
+        long bookId = 12L;
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java:64]: You can move it to the constants with informative name

File: src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        MvcResult result = mockMvc.perform(get("/books")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+        BookResponseDto[] actual = objectMapper.readValue(result.getResponse()
+                        .getContentAsByteArray(),
+                BookResponseDto[].class);
+
+        Assertions.assertEquals(2, actual.length);
+        Assertions.assertEquals(expected, Arrays.stream(actual).toList());
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java:122]:
```suggestion
        Assertions.assertEquals(expected, actual);
```

File: src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    void findAllBooks_BooksExistInCatalog_ShouldReturnAllBooks_Ok() throws Exception {
+        List<BookResponseDto> expected = createBooksList();
+
+        MvcResult result = mockMvc.perform(get("/books")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+        BookResponseDto[] actual = objectMapper.readValue(result.getResponse()
+                        .getContentAsByteArray(),
+                BookResponseDto[].class);
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java:119]:
```suggestion
        List<BookResponseDto> actual = objectMapper.readValue(result.getResponse().getContentAsString(), new TypeReference<>() {});
```

File: src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+                .setTitle(requestDto.getTitle())
+                .setAuthor(requestDto.getAuthor())
+                .setDescription(requestDto.getDescription())
+                .setIsbn(requestDto.getIsbn())
+                .setPrice(requestDto.getPrice())
+                .setCoverImage(requestDto.getCoverImage())
+                .setCategoryIds(requestDto.getCategoryIds());
+    }
+
+    private List<BookResponseDto> createBooksList() {
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java:188]: You can use `List.of()` method

File: src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        return new BookResponseDto()
+                .setId(id)
+                .setTitle("Test Book 1")
+                .setAuthor("Test Author 1")
+                .setDescription("Description for Test Book 1")
+                .setIsbn("9783161484100")
+                .setPrice(BigDecimal.valueOf(19.99))
+                .setCoverImage("http://example.com/cover1.jpg")
+                .setCategoryIds(Set.of(12L));
+    }
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/BookControllerTest.java:221]: You can move these methods to separate `TestUtils` class

File: src/test/java/com/onlinebookstore/store/controllertests/CategoryControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@Sql(scripts = "classpath:database/test/categories/add-categories-to-table.sql",
+        executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+@Sql(scripts = "classpath:database/test/books/delete-books-categories.sql",
+        executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/CategoryControllerTest.java:33]:
```suggestion
```

File: src/test/java/com/onlinebookstore/store/controllertests/CategoryControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+        CategoryResponseDto actualDto = objectMapper.readValue(result.getResponse()
+                        .getContentAsString(),
+                CategoryResponseDto.class);
+        Assertions.assertEquals(createCategoryDto
+                        .getName(),
+                actualDto.getName());
+        Assertions.assertEquals(createCategoryDto
+                        .getDescription(),
+                actualDto.getDescription());
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/CategoryControllerTest.java:79]: Use static import
```suggestion
        assertEquals(createCategoryDto.getName(), actualDto.getName());
        assertEquals(createCategoryDto.getDescription(), actualDto.getDescription());
```

File: src/test/java/com/onlinebookstore/store/controllertests/CategoryControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    void getAllCategories_ShouldReturnCategoriesList_Ok() throws Exception {
+        MvcResult result = mockMvc.perform(get("/categories")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        CategoryResponseDto[] actual = objectMapper.readValue(result.getResponse()
+                        .getContentAsString(),
+                CategoryResponseDto[].class);
+        Assertions.assertTrue(actual.length > 0);
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/CategoryControllerTest.java:94]: Will be better to check if we have result with exactly the same size as we expect

File: src/test/java/com/onlinebookstore/store/controllertests/CategoryControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    void updateCategory_ValidId_ShouldReturnUpdatedCategory_Ok() throws Exception {
+        long categoryId = 12L;
+        CategoryResponseDto updateCategoryDto = new CategoryResponseDto()
+                .setName("Updated Category")
+                .setDescription("Updated description");
+
+        String jsonRequest = objectMapper.writeValueAsString(updateCategoryDto);
+
+        MvcResult result = mockMvc.perform(
+                        put("/categories/" + categoryId)
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/CategoryControllerTest.java:109]:
```suggestion
                        put("/categories/{id}", categoryId)
```

File: src/test/java/com/onlinebookstore/store/controllertests/CategoryControllerTest.java [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+                        .getContentAsString(),
+                CategoryResponseDto[].class);
+        Assertions.assertTrue(actual.length > 0);
+    }
+
+    @Test
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
+    @DisplayName("Update Category")
+    void updateCategory_ValidId_ShouldReturnUpdatedCategory_Ok() throws Exception {
+        long categoryId = 12L;
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/CategoryControllerTest.java:101]: Move to the contants

File: src/test/java/com/onlinebookstore/store/controllertests/CategoryControllerTest.java [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+        BookDtoWithoutCategoryIds bookDto2 = new BookDtoWithoutCategoryIds()
+                .setTitle("Book 2")
+                .setAuthor("Author 2")
+                .setPrice(new BigDecimal("12.99"))
+                .setIsbn("9783161484101")
+                .setDescription("Description 2")
+                .setCoverImage("http://example.com/cover2.jpg");
+
+        return List.of(bookDto1, bookDto2);
+    }
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/CategoryControllerTest.java:148]: You can move these methods to separate `TestUtils` class

File: src/test/java/com/onlinebookstore/store/controllertests/CategoryControllerTest.java [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+
+    private CategoryResponseDto createTestCategoryResponseDto() {
+        return new CategoryResponseDto()
+                .setId(1L)
+                .setName("Test Category")
+                .setDescription("Test description");
+    }
+
+    private List<BookDtoWithoutCategoryIds> createTestBookDtoList() {
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/controllertests/CategoryControllerTest.java:130]: Seems unused

File: src/test/java/com/onlinebookstore/store/repositorytests/BookRepositoryTest.java [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+
+        Book secondBook = new Book();
+        secondBook.setId(13L);
+        secondBook.setTitle("Test Book 2");
+        secondBook.setAuthor("Test Author 2");
+        secondBook.setIsbn("9783161484101");
+        secondBook.setPrice(BigDecimal.valueOf(29.99));
+        secondBook.setDescription("Description for Test Book 2");
+        secondBook.setCoverImage("http://example.com/cover2.jpg");
+        secondBook.setCategories(Set.of(category));
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/repositorytests/BookRepositoryTest.java:64]: Move to the separate methods

File: src/test/java/com/onlinebookstore/store/servicetests/BookServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+        assertEquals(expectedBookResponseDto, bookResponseDto);
+    }
+
+    @Test
+    @DisplayName("Finds books by id")
+    void findsBook_ById_ShouldThrowException_NotOk() {
+        when(bookRepository
+                .findBookById(anyLong()))
+                .thenReturn(Optional.empty());
+        assertThrows(EntityNotFoundException.class,
```
Comment (JJJazl) [inline src/test/java/com/onlinebookstore/store/servicetests/BookServiceTest.java:68]: Check error message content

**Review by JJJazl — CHANGES_REQUESTED** [review]

**Review by okuzan — APPROVED** [review]
